components/script/dom/node.rs

```diff
@@ -2726,18 +2779,109 @@ impl NodeMethods for Node {
     }
 
     // https://dom.spec.whatwg.org/#dom-node-comparedocumentposition
-    fn CompareDocumentPosition(&self, other: &Node) -> u16 {
+    fn CompareDocumentPosition(&self, other: Option<&Node>) -> u16 {
         // step 1.
-        if self == other {
+        if Some(self) == other {
             return 0;
         }
 
-        // FIXME(emilio): This will eventually need to handle attribute nodes.
+        // step 2
+        let mut node1 = other;
+        let mut node2 = Some(self);
+
+        // step 3
+        let mut attr1: Option<&Attr> = None;
+        let mut attr2: Option<&Attr> = None;
+
+        // step 4
+        // The compiler doesn't know the lifetime of attr1.GetOwnerElement
+        // is guaranteed by the lifetime of attr1, so we hold it explicitly
+        let attr1owner;
+        match node1 {
+            Some(ref n) => {
+                if let Some(ref a) = n.downcast::<Attr>() {
+                    attr1 = Some(a);
+                    attr1owner = a.GetOwnerElement();
+                    node1 = match attr1owner {
+                        Some(ref e) => Some(&e.upcast()),
+                        None => None,
+                    }
+                }
+            },
+            None => {},
+        };
 
-        let mut self_and_ancestors = self
+        // step 5.1: spec says to operate on node2 here,
+        // node2 is definitely just Some(self) going into this step
+        let attr2owner;
+        if let Some(ref a) = self.downcast::<Attr>() {
+            attr2 = Some(a);
+            attr2owner = a.GetOwnerElement();
+            node2 = match attr2owner {
+                Some(ref e) => Some(&*e.upcast()),
+                None => None,
+            }
+        }
+
+        // Step 5.2
+        // This substep seems lacking in test coverage.
+        // We hit this when comparing two attributes that have the
+        // same owner element.
+        if !node1.is_none() && node1 == node2 {
```

Comment:
wrap the whole thing in `if let Some(node1) = node1` so you don't have to unwrap

Reply:
(also, for future reference, `!is_none` is `is_some()`)

Reply:
Will do!